oci-image/src/pull.rs

```diff
@@ -0,0 +1,188 @@
+    pub fn new(image: &str, image_dir: &Path) -> Result<Self> {
+        let reference = Reference::try_from(image)
+            .map_err(|_| Error::InvalidOCIReference(image.to_string()))?;
+        let auth = RegistryAuth::Anonymous;
+
+        Ok(Self {
+            auth,
+            reference,
+            client: Client::default(),
+            image_dir: image_dir.to_path_buf(),
+            accepted_media_types: vec![],
+        })
+    }
+
+    /// Pull the image manifest and the configuration from the registry.
+    pub async fn pull_manifest(&mut self) -> Result<(OciManifest, String, String)> {
+        let (manifest, digest, config) = self
+            .client
+            .pull_manifest_and_config(&self.reference, &self.auth)
+            .await
+            .map_err(|e| Error::PullManifest(e.to_string()))?;
+
+        // Build a list of the accepted media types
+        self.accepted_media_types = manifest
+            .layers
+            .iter()
+            .map(|x| x.media_type.clone())
+            .collect::<Vec<String>>();
+
+        // Remove duplicates
+        self.accepted_media_types.dedup();
+
+        Ok((manifest, digest, config))
+    }
+
+    /// Pull the image from the registry. The method will return the complete path to the image on disk.
+    ///
+    /// The image will be stored into the `image_dir` argument provided in `Puller` constructor.
+    /// This will produce a fully compliant OCI image with a `blobs` directory and a `manifest.json` file.
+    pub async fn pull_layers(
+        &mut self,
+        state: Arc<Mutex<State>>,
+        diffs: &[String],
+    ) -> Result<Vec<LayerMetadata>> {
+        // Get the image data
+        let data = self
+            .client
+            .pull(
+                &self.reference,
+                &self.auth,
+                Vec::from_iter(self.accepted_media_types.iter().map(String::as_str)),
+            )
+            .await
+            .map_err(|e| Error::PullImage(e.to_string()))?;
+
+        let layers_dir = self.image_dir.join("layers");
+        // Create the directories to store layers if not exists
+        create_dir_all(&layers_dir).map_err(|e| Error::LayerDirectoryCreation(e.to_string()))?;
+
+        let layers = data.layers.into_iter().enumerate().map(|(i, layer)| {
+            let state = state.clone();
+            let layer_data = layer.data.clone();
+            let layer_sha256_digest = layer.sha256_digest();
+            let layer_path = layers_dir.join(&layer_sha256_digest.replace(':', "_"));
+            // This block defines the behavior for one layer.
+            async move {
+                // If the layer already exists on the disk, return directly
+                if let Some(layer_meta) = state.lock().await.get_layer(&layer_sha256_digest) {
+                    log::info!("Layer {} : {} (cached)", &i, &layer_sha256_digest);
+                    return Ok::<_, crate::Error>(layer_meta.clone());
+                }
```

Comment:
This could be moved before the `async move` and avoid a whole `state` clone when the layer already exists.

Reply:
I think it's not possible because `state` needs to be locked. As I use the `tokio::sync::Mutex`, I have to use await and in consequence be in an async block?

Reply:
ah of course...tokio. Fine with me then.